src/librustc_mir/interpret/visitor.rs

```diff
@@ -142,6 +142,7 @@ macro_rules! make_value_visitor {
                 self.walk_value(v)
             }
             /// Visit the given value as a union.  No automatic recursion can happen here.
+            /// Also called for the fields of a generator, which may or may not be initialized.
```

Comment:
I don't see this happening in the code below.

Reply:
Oh yeah I went back on this because it dosn't work very well... I guess I could still do it an go through `visit_field` though.

Reply:
Actually no that doesn't work, it doesn't have a union type. I don't think there is a way to visit the other generator fields at all with the current interface, and it doesn't seem worth extending the interface?

Reply:
Oh yea, that's totally fine, as long as the comments mirror reality ;)

Well, as long as validation doesn't get hickups elsewhere because https://github.com/solson/miri/blob/adfede5cec2c8a136830f7fc309dbb45ac7a098a/src/helpers.rs#L221 wasn't visited in miri.

Reply:
Hm, that is a good point. I forgot that I was using `visit_union` there.

This is relevant when determining where there are `UnsafeCell` inside a generator. If there is no `UnsafeCell`, shared references enforce memory to be frozen. So we *probably* should go conservatively type-based here like we do for unions... dang.

Just calling `visit_union` after doing the field projections would actually work, but it would violate the protocol that lets a visitor keep track of which "path" inside the data structure we are at. The only visitor relying on the path is validation, which doesn't do anything for unions, so this is fine in principle... but it's not nice. Any ideas?

Reply:
I added a new `visit_generator_field` hook for this. Now at least it makes sense, and likely nobody will ever overwrite that hook...